### PR TITLE
Updated compress and decompress functions

### PR DESCRIPTION
**Pull request by dishvyas**
Added functions in utils.py and replaced other occurences with function calls from utils.py

### CODE REVIEW

**Comment by marco-c** [comment]
@dishvyas you don't need to create new PRs, you can just update your existing one (#780).

**Comment by dishvyas** [comment]
> @dishvyas you don't need to create new PRs, you can just update your existing one (#780).

Sorry I'm just new at this. And i had trouble getting it right. Are the changes right at all?

**Comment by marco-c** [comment]
> Sorry I'm just new at this. And i had trouble getting it right. Are the changes right at all?

They are more or less right (with some changes needed), but you have another PR open for this, you can update that one instead of creating new ones :)

**Comment by dishvyas** [comment]
> > Sorry I'm just new at this. And i had trouble getting it right. Are the changes right at all?
> 
> They are more or less right (with some changes needed), but you have another PR open for this, you can update that one instead of creating new ones :)

Do i need to make more changes before i add it to previous PR?

**Comment by marco-c** [comment]
> Do i need to make more changes before i add it to previous PR?

Please update your other PR, and I will do the review there. It's better to avoid duplicating the review comments in multiple PRs.

If you have troubles updating that PR for some reason, just let me know, I'll close that one and you can continue your work on this one.

**Comment by dishvyas** [comment]
> > Do i need to make more changes before i add it to previous PR?
> 
> Please update your other PR, and I will do the review there. It's better to avoid duplicating the review comments in multiple PRs.
> 
> If you have troubles updating that PR for some reason, just let me know, I'll close that one and you can continue your work on this one.

Sorry i can't seem to find a way to update the previous one. If you could do that, it would be great! :)

**Comment by marco-c** [comment]
Done!

OK, so the first thing to fix is the linting issues (you can see them on Taskcluster by clicking on the "Details" link). As I mentioned in the other PR, you should be able to use pre-commit to fix them semi-automatically.

File: bugbug/utils.py [hunk 1]
```diff
@@ -19,6 +19,20 @@
 TASKCLUSTER_DEFAULT_URL = "https://taskcluster.net"
 
 
+def compress(path):
```
Comment (marco-c) [inline bugbug/utils.py]: Call the two functions zstd_compress and zstd_decompress, since they are specifically about zstandard.

File: bugbug/utils.py [hunk 2]
```diff
@@ -162,10 +176,10 @@ class CustomJsonEncoder(json.JSONEncoder):
     """ A custom Json Encoder to support Numpy types
     """
 
-    def default(self, obj):
+    def default(self, obj):          # pylint: disable=E0202
```
Comment (marco-c) [inline bugbug/utils.py]: This is an unrelated change, please remove it.

File: bugbug/utils.py [hunk 3]
```diff
@@ -162,10 +176,10 @@ class CustomJsonEncoder(json.JSONEncoder):
...
         try:
             return np.asscalar(obj)
         except (ValueError, IndexError, AttributeError, TypeError):
             pass
 
-        return super().default(self, obj)
+        return json.JSONEncoder.default(self, obj)
```
Comment (marco-c) [inline bugbug/utils.py]: This is an unrelated change, please remove it.

File: bugbug/utils.py [hunk 4]
```diff
@@ -19,6 +19,20 @@
...
+    cctx = zstandard.ZstdCompressor()
+    with open(path, "rb") as input_f:
+        with open(f"{path}.zst", "wb") as output_f:
+            cctx.copy_stream(input_f, output_f)
+
+
+        
+def decompress(path):
+    dctx = zstandard.ZstdDecompressor()
+    with open(path, "rb") as input_f:
```
Comment (marco-c) [inline bugbug/utils.py]: You're reading from the file which has to be extracted and writing to the file from which we should extract, I guess you got this backwards :)

I suggest testing the functions yourself manually with a Python script (write a file, compress it, remove the original file, decompress the archive, check that the output file contains the same content you wrote at the beginning).

File: http_service/models.py [hunk 5]
```diff
@@ -14,7 +14,7 @@
 
 from bugbug import bugzilla, get_bugbug_version
 from bugbug.models import load_model
-
+from bugbug.utils import decompress
```
Comment (marco-c) [inline http_service/models.py]: Do `from bugbug import utils`, and then call utils.decompress

File: scripts/check.py [hunk 6]
```diff
@@ -31,7 +28,7 @@ def go(self, model_name):
         download_url = os.getenv("MODEL_DOWNLOAD_URL")
 
         if should_download_model and download_url:
-            download_url = download_model(download_url, f"{model_name}model")
+            download_model(download_url, f"{model_name}model")
```
Comment (marco-c) [inline scripts/check.py]: This is an unrelated change, please revert it

File: scripts/commit_classifier.py [hunk 7]
```diff
@@ -29,10 +29,7 @@ def __init__(self, cache_root):
 
         if not os.path.exists("regressormodel"):
             download_check_etag(URL, "regressormodel.zst")
-            dctx = zstandard.ZstdDecompressor()
-            with open("regressormodel.zst", "rb") as input_f:
-                with open("regressormodel", "wb") as output_f:
-                    dctx.copy_stream(input_f, output_f)
+            decompress("regressormodel.zst")
```
Comment (marco-c) [inline scripts/commit_classifier.py]: In other places you are passing the path without `zst`, here you are passing the path with `zst`. One of the two is wrong

File: scripts/regressor_finder.py [hunk 8]
```diff
@@ -402,9 +395,7 @@ def find_bic(bug_fixing_commit):
 
             return bug_introducing_commits
 
-        with concurrent.futures.ThreadPoolExecutor(
-            initializer=_init, initargs=(repo_dir,), max_workers=os.cpu_count() + 1
-        ) as executor:
+        with concurrent.futures.ThreadPoolExecutor(initializer=_init, initargs=(repo_dir,), max_workers=os.cpu_count() + 1) as executor:
```
Comment (marco-c) [inline scripts/regressor_finder.py]: Unrelated change

**Comment by dishvyas** [comment]
@marco-c  Can you check the latest functions i have written and help me if i need to make any more changes??

**Comment by marco-c** [comment]
@dishvyas I'm sorry but the PR is unreviewable in this state, take a look at the changes for yourself: https://github.com/mozilla/bugbug/pull/790/files.
You have a lot of unresolved merge conflicts, duplicated functions, commented-out code...

File: bugbug/utils.py (hunk 2 above)
Comment (marco-c) [reply bugbug/utils.py]: This comment still applies.

File: bugbug/utils.py (hunk 3 above)
Comment (marco-c) [reply bugbug/utils.py]: This comment still applies.

File: bugbug/utils.py [hunk 9]
```diff
@@ -18,6 +18,27 @@
 
 TASKCLUSTER_DEFAULT_URL = "https://taskcluster.net"
 
+def zstd_compress(path):
```
Comment (marco-c) [inline bugbug/utils.py]: You should just use copy_stream, copying the preexisting code, why are you complicating your life?

File: bugbug/db.py [hunk 10]
```diff
@@ -157,15 +146,9 @@ def _db_open(path, mode):
             yield store_constructor(f)
     elif compression == "zstd":
         if "w" in mode or "a" in mode:
-            cctx = zstandard.ZstdCompressor()
-            with open(path, mode) as f:
-                with cctx.stream_writer(f) as writer:
-                    yield store_constructor(writer)
+            utils.zstd_compress(path)
```
Comment (marco-c) [inline bugbug/db.py]: This is wrong, the old code was not compressing a file. In this case, keep the old code.

File: bugbug/db.py [hunk 11]
```diff
@@ -157,15 +146,9 @@ def _db_open(path, mode):
...
         else:
-            dctx = zstandard.ZstdDecompressor()
-            with open(path, mode) as f:
-                with dctx.stream_reader(f) as reader:
-                    yield store_constructor(reader)
+            utils.zstd_decompress(path)
```
Comment (marco-c) [inline bugbug/db.py]: This is wrong, the old code was not decompressing a file. In this case, keep the old code.

File: bugbug/db.py [hunk 12]
```diff
@@ -60,7 +49,7 @@ def download_support_file(path, file_name):
         utils.download_check_etag(url, path)
 
         if path.endswith(".zst"):
-            extract_file(path)
+            utils.zstd_decompress(path)
```
Comment (marco-c) [inline bugbug/db.py]: You are passing a path with the `.zst` extension here, and without the extension at https://github.com/mozilla/bugbug/pull/790/files#diff-2f714e5a8f0bf3c9fa77453ee08b8ddaR79.

**Comment by dishvyas** [comment]
> @dishvyas I'm sorry but the PR is unreviewable in this state, take a look at the changes for yourself: https://github.com/mozilla/bugbug/pull/790/files.
> You have a lot of unresolved merge conflicts, duplicated functions, commented-out code...
I have tried linting using pytest and therefore i was trying to install all the modules from requirements.txt. After that I'm getting this error and i tried googling it but couldn't find anything like 'jsone' module.
EModuleNotFoundError: No module named 'jsone'

**Comment by marco-c** [comment]
`pre-commit` is used for linting, I told you many times now :P
`pytest` is used for testing.

> After that I'm getting this error and i tried googling it but couldn't find anything like 'jsone' module.
EModuleNotFoundError: No module named 'jsone'

https://pypi.org/project/json-e/

You need to install requirements from all requirement files:
```
pip install --quiet -r requirements.txt
pip install --quiet -r extra-nlp-requirements.txt
pip install --quiet -r extra-nn-requirements.txt
pip install --quiet -r infra/spawn_data_pipeline_requirements.txt
pip install --quiet -r test-requirements.txt
```

File: bugbug/utils.py (hunk 9 above)
Comment (dishvyas) [reply bugbug/utils.py]: Are you just talking about the compress_file code or decompress one too?? Because decompress_file is not exactly decompressing the file when i tried running it on a seperate python script to test it.

Comment (marco-c) [reply bugbug/utils.py]: I'm talking about both, the goal of this PR is just to move the functions in utils.py, you shouldn't need to change anything inside the functions themselves.

Comment (dishvyas) [reply bugbug/utils.py]: @marco-c Alright my bad! One last thing since when i first made the commit you told me to check if the functions are running or not, i tried testing them on a file. Compression was successful but I'm not sure if the decompress function was working so let me know if this is the correct one for zstd_decompress:

def zstd_decompress(path):
    dctx = zstandard.ZstdDecompressor()
    with open(f"{path}.zst", "rb") as input_f:
        with open(path, "wb") as output_f:
            dctx.copy_stream(input_f, output_f)

Comment (marco-c) [reply bugbug/utils.py]: > I'm not sure if the decompress function was working

You can simply run your function to make sure it works.

**Comment by dishvyas** [comment]
@marco-c  Do i need to stage the .pre-commit-config.yaml file in the commit??

**Comment by marco-c** [comment]
> @marco-c Do i need to stage the .pre-commit-config.yaml file in the commit??

The pre-commit configuration is unrelated to this PR, so you should not change it.

@dishvyas sorry, but I think this is not a good way to go forward, you should try to get some experience with Python and git working on some personal project first.
